Disable MESS

meowsbits opened the pull request:
This PR drafts the disablement of  [MESS/ECBP1100](https://ecips.ethereumclassic.org/ECIPs/ecip-1100), anticipated to occur on ETC concurrently with the [Spiral](https://ecips.ethereumclassic.org/ECIPs/ecip-1109) network upgrade. 

Following convention, a configuration override flag `--ecbp1100.disable=[number]` is provided.

As-is, this draft uses `runtime` and `reflect` at the `Configurator` implementation level to handle the actual enable/disable logic (https://github.com/etclabscore/core-geth/commit/9e274d02851888a9dab51d5ba197193f09aa2b24). Alternatively, we could consider instead adding conditions at the places where the configuration is referenced to handle it, like the example below.

```diff
diff --git a/core/forkchoice.go b/core/forkchoice.go
index d07cdf6aa4..55d898c3a0 100644
--- a/core/forkchoice.go
+++ b/core/forkchoice.go
@@ -162,7 +162,8 @@ func (f *ForkChoice) ReorgNeeded(current *types.Header, extern *types.Header) (b
                        return reorg, nil
                }
        }
-       if !f.chain.Config().IsEnabled(f.chain.Config().GetECBP1100Transition, current.Number) {
+       if !f.chain.Config().IsEnabled(f.chain.Config().GetECBP1100Transition, current.Number) ||
+               f.chain.Config().IsEnabled(f.chain.Config().GetECBP1100DisableTransition, current.Number) {
                return reorg, nil
        }

```

meowsbits committed:
cmd/geth,cmd/utils,eth,eth/ethconfig,params/types/coregeth,params/types/ctypes,params/types/genesisT,params/types/goethereum: ECBP1100Disable: cli flag for override, config implementation

Date: 2023-11-29 14:45:34-07:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
params: configure ETC's ECBP1100Disable with FAKE block number

Date: 2023-11-29 14:46:29-07:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
params/types/coregeth,params/types/goethereum: implement ECBP1100Disable logic with runtime,reflection

This may not be ideal from a CPU/time
perspective (expected slow), but since
this is likely a temporary code path
as we expect to remove MESS code once
the fork is passed, it may be sufficient,
and is certainly a smaller change than
adding -Disabled conditions at each
occurrence of IsEnabled/ECBP1100 in the code.

Date: 2023-11-29 15:52:37-07:00
Signed-off-by: meows <b5c6@protonmail.com>

ziogaschr commented on params/types/coregeth/chain_config_configurator.go:
```diff
@@ -669,6 +681,13 @@ func (c *CoreGethChainConfig) IsEnabled(fn func() *uint64, n *big.Int) bool {
 	if f == nil || n == nil {
 		return false
 	}
+	fnName := runtime.FuncForPC(reflect.ValueOf(fn).Pointer()).Name()
```
Nice hack in order to check if both are enabled, either way it will not stay for too long.

p.s.: I hope I don't read back this comment in 1-2 years and the code is still here :)

ziogaschr committed:
cmd/utils,core: return a warning if both --ecbp1100.disable and --ecbp1100.nodisable

ziogaschr commented on cmd/utils/flags.go:
```diff
@@ -2081,6 +2086,8 @@ func SetEthConfig(ctx *cli.Context, stack *node.Node, cfg *ethconfig.Config) {
 			cfg.EthDiscoveryURLs = SplitAndTrim(urls)
 		}
 	}
+	CheckExclusive(ctx, ECBP1100DisableFlag, ECBP1100NoDisableFlag)
```
@meowsbits this is probably better to be removed (I added it).

A node operator might indeed want the `nodisable` functionality, so if he still wants to enable ECBP-1100 he has to do `--ecbp1100.disable=69420123 --ecbp1100.nodisable`.

The thing is that with the above check, this is not allowed.

diega commented on params/config_classic.go:
```diff
@@ -81,8 +81,9 @@ var (
-		ECBP1100FBlock: big.NewInt(11_380_000), // ETA 09 Oct 2020
-		ECIP1099FBlock: big.NewInt(11_700_000), // Etchash (DAG size limit)
+		ECBP1100FBlock:        big.NewInt(11_380_000), // ETA 09 Oct 2020
+		ECBP1100DisableFBlock: big.NewInt(69_420_123),
```
What does `69_420_123` mean?

ziogaschr replied:
Just far in future. I will not be surprised if it has a meaning though :)

meowsbits replied:
Yea its just a placeholder block number.

meowsbits replied:
Configured to == Spiral w/ https://github.com/etclabscore/core-geth/pull/592/commits/907295dcdbcc66bf4474eae22f3341c0941ea3b9

diega reviewed (approved):
🚢

meowsbits committed:
Merge branch 'master' into disable-mess

meowsbits committed:
cmd/utils: these flags are not exclusive

- 1100disable flag deactivates the feature
- 1100.nodisable prevents safety mechanisms from
  toggling it once activated

Date: 2023-12-06 07:11:04-07:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
core,eth,params/types/coregeth,params/types/ctypes,params/types/genesisT,params/types/goethereum: s/ECBP1100DisableTransition/ECBP1100DeactivateTransition/g

The word Disable conflicts with the existing
ECBP1100NoDisable flag and feature.
That feature prevents the toggling of the ECBP1100
feature per the safety mechanisms of low peer count
and stale head.

Renaming to Deactivate is intended to clarify
the code usages.

Based on feedback provided here https://github.com/etclabscore/core-geth/pull/592#issuecomment-1841126001

Date: 2023-12-06 07:21:58-07:00
Signed-off-by: meows <b5c6@protonmail.com>

cmd/geth,cmd/utils,core,eth,eth/ethconfig,params/types/coregeth: ECBP1100: s/enable/activate/g,s/disable/deactivate/g

rename all references besides the method name

Date: 2023-12-06 07:50:14-07:00
Signed-off-by: meows <b5c6@protonmail.com>

params,params/types/coregeth,params/types/goethereum: ECBP1100-deactivate: rename config fields s/disable/deactivate/g

Date: 2023-12-06 08:07:08-07:00
Signed-off-by: meows <b5c6@protonmail.com>

core: (lint) remove unnecessary leading newline

Date: 2023-12-06 08:09:00-07:00
Signed-off-by: meows <b5c6@protonmail.com>

params/types/goethereum: ECBP100: s/disableTransition/deactivateTransition/g

Date: 2023-12-06 08:17:12-07:00
Signed-off-by: meows <b5c6@protonmail.com>

params/types/coregeth: ECBP100: s/disable/deactivate/g

Date: 2023-12-06 08:19:09-07:00
Signed-off-by: meows <b5c6@protonmail.com>

ziogaschr commented on params/types/coregeth/chain_config.go:
```diff
@@ -184,8 +184,9 @@ type CoreGethChainConfig struct {
-	ECBP1100FBlock *big.Int `json:"ecbp1100FBlock,omitempty"` // ECBP1100:MESS artificial finality
+	ECIP1099FBlock        *big.Int `json:"ecip1099FBlock,omitempty"`        // ECIP1099 etchash HF block
+	ECBP1100FBlock        *big.Int `json:"ecbp1100FBlock,omitempty"`        // ECBP1100:MESS artificial finality
+	ECBP1100DisableFBlock *big.Int `json:"ecbp1100DisableFBlock,omitempty"` // Disable ECBP1100:MESS artificial finality
```
```suggestion
	ECBP1100DeactivateFBlock *big.Int `json:"ecbp1100DeactivateFBlock,omitempty"` // Deactivate ECBP1100:MESS artificial finality
```

meowsbits replied:
Done and squashed into https://github.com/etclabscore/core-geth/pull/592/commits/e4b7dd2f1b2ebcfe7c62f29e22bc4eebfee1881d

meowsbits committed:
cmd/geth,cmd/utils,eth,eth/ethconfig: ECBP100-deactivate: move override flags to --override.X,

and categorize these flags.

Date: 2023-12-06 08:02:58-07:00
Signed-off-by: meows <b5c6@protonmail.com>

ziogaschr commented on params/config_classic.go:
```diff
@@ -81,8 +81,9 @@ var (
-		ECBP1100FBlock: big.NewInt(11_380_000), // ETA 09 Oct 2020
-		ECIP1099FBlock: big.NewInt(11_700_000), // Etchash (DAG size limit)
+		ECBP1100FBlock:           big.NewInt(11_380_000), // ETA 09 Oct 2020
+		ECBP1100DeactivateFBlock: big.NewInt(69_420_123),
```
Let's not forget to set the block_number here or we wait for this PR to be merged and set it at #595

meowsbits replied:
I just set it here == Spiral https://github.com/etclabscore/core-geth/pull/592/commits/907295dcdbcc66bf4474eae22f3341c0941ea3b9

ziogaschr commented on params/types/coregeth/chain_config.go:
```diff
@@ -184,8 +184,9 @@ type CoreGethChainConfig struct {
-	ECBP1100FBlock *big.Int `json:"ecbp1100FBlock,omitempty"` // ECBP1100:MESS artificial finality
+	ECIP1099FBlock           *big.Int `json:"ecip1099FBlock,omitempty"`                 // ECIP1099 etchash HF block
+	ECBP1100FBlock           *big.Int `json:"ecbp1100FBlock,omitempty"`                 // ECBP1100:MESS artificial finality
+	ECBP1100DeactivateFBlock *big.Int `json:"ecbp1100DeactivateFBlockFBlock,omitempty"` // Disable ECBP1100:MESS artificial finality
```
```suggestion
	ECBP1100DeactivateFBlock *big.Int `json:"ecbp1100DeactivateFBlockFBlock,omitempty"` // Deactivate ECBP1100:MESS artificial finality
```

meowsbits replied:
I made it and squashed into https://github.com/etclabscore/core-geth/pull/592/commits/e4b7dd2f1b2ebcfe7c62f29e22bc4eebfee1881d

ziogaschr commented on params/types/goethereum/goethereum_configurator.go:
```diff
@@ -691,6 +703,13 @@ func (c *ChainConfig) IsEnabled(fn func() *uint64, n *big.Int) bool {
 	}
+	fnName := runtime.FuncForPC(reflect.ValueOf(fn).Pointer()).Name()
+	if strings.Contains(fnName, "ECBP1100Transition") {
+		disabledTransition := c.GetECBP1100DeactivateTransition()
```
@meowsbits  this is the last one. `s/disabledTransition/deactivateTransition/g`

meowsbits replied:
Danke, done and force pushed again (squashed into https://github.com/etclabscore/core-geth/pull/592/commits/7a026c079bf33bceca4629cb8132628f99baad41)

meowsbits committed:
params: ECBP1100(MESS): configure deactivation 19_250_000 == Spiral

This configures the deactivation of MESS
to coincide with the Spiral hardfork.

Date: 2023-12-06 08:29:02-07:00
Signed-off-by: meows <b5c6@protonmail.com>

diega reviewed (approved):
🚢²
